lib/datastore/entity.js

```diff
@@ -378,63 +396,79 @@ module.exports.propertyToValue = propertyToValue;
  * //   stringValue: 'Hi'
  * // }
  */
-function valueToProperty(v) {
-  var p = {};
-  if (v instanceof Boolean || typeof v === 'boolean') {
-    p.boolean_value = v;
-    return p;
-  }
-  if (v instanceof Int) {
-    p.integer_value = v.get();
-    return p;
-  }
-  if (v instanceof Double) {
-    p.double_value = v.get();
-    return p;
+function valueToProperty(value) {
+  var propertyObject = {};
+
+  if (is.boolean(value)) {
+    propertyObject.booleanValue = value;
+    return propertyObject;
   }
-  if (v instanceof Number || typeof v === 'number') {
-    if (v % 1 === 0) {
-      p.integer_value = v;
+
+  if (is.number(value)) {
+    if (value % 1 === 0) {
+      value = new Int(value);
     } else {
-      p.double_value = v;
+      value = new Double(value);
     }
-    return p;
   }
-  if (v instanceof Date) {
-    p.timestamp_microseconds_value = v.getTime() * 1000;
-    return p;
+
+  if (value instanceof Int) {
+    propertyObject.integerValue = value.get();
+    return propertyObject;
   }
-  if (v instanceof String || typeof v === 'string') {
-    p.string_value = v;
-    return p;
+
+  if (value instanceof Double) {
+    propertyObject.doubleValue = value.get();
+    return propertyObject;
   }
-  if (v instanceof Buffer) {
-    p.blob_value = v;
-    return p;
+
+  if (value instanceof Date) {
+    propertyObject.timestampValue = {
+      seconds: value.getTime() / 1000
+    };
+    return propertyObject;
   }
-  if (Array.isArray(v)) {
-    p.list_value = v.map(function(item) {
-      return valueToProperty(item);
-    });
-    return p;
+
+  if (is.string(value)) {
+    propertyObject.stringValue = value;
+    return propertyObject;
+  }
+
+  if (value instanceof Buffer) {
+    propertyObject.blobValue = value;
+    return propertyObject;
+  }
+
+  if (is.array(value)) {
+    propertyObject.arrayValue = {
+      values: value.map(valueToProperty)
+    };
+    return propertyObject;
   }
-  if (v instanceof Key) {
-    p.key_value = keyToKeyProto(v);
-    return p;
+
+  if (value instanceof Key) {
+    propertyObject.keyValue = keyToKeyProto(value);
+    return propertyObject;
   }
-  if (v instanceof Object && Object.keys(v).length > 0) {
-    var property = [];
-    Object.keys(v).forEach(function(k) {
-      property.push({
-        name: k,
-        value: valueToProperty(v[k])
-      });
+
+  if (is.object(value) && Object.keys(value).length > 0) {
+    var property = Object.keys(value).map(function(key) {
+      return {
+        name: key,
+        value: valueToProperty(value[key])
```

Comment:
Note that is now acceptable for these values to be indexes (i.e. if I put an object:

    {
      a: {
        b: 2
      }
    }

Then I can actually query for this using the property `a.b`.